src/common/datasource/src/compression.rs

```diff
@@ -72,38 +73,98 @@ impl CompressionType {
     pub const fn is_compressed(&self) -> bool {
         !matches!(self, &Self::UNCOMPRESSED)
     }
+}
 
-    pub fn convert_async_read<T: AsyncRead + Unpin + Send + 'static>(
-        &self,
-        s: T,
-    ) -> Box<dyn AsyncRead + Unpin + Send> {
-        match self {
-            CompressionType::GZIP => Box::new(GzipDecoder::new(BufReader::new(s))),
-            CompressionType::BZIP2 => Box::new(BzDecoder::new(BufReader::new(s))),
-            CompressionType::XZ => Box::new(XzDecoder::new(BufReader::new(s))),
-            CompressionType::ZSTD => Box::new(ZstdDecoder::new(BufReader::new(s))),
-            CompressionType::UNCOMPRESSED => Box::new(s),
-        }
-    }
+macro_rules! impl_compression_type {
+    ($(($enum_item:ident, $prefix:ident)),*) => {
+        paste::item! {
+            impl CompressionType {
+                pub async fn encode(&self, content: impl AsRef<[u8]>) -> io::Result<Vec<u8>> {
+                    match self {
+                        $(
+                            CompressionType::$enum_item => {
+                                let mut buffer = Vec::new();
+                                let mut encoder = write::[<$prefix Encoder>]::new(&mut buffer);
+                                encoder.write_all(content.as_ref()).await?;
+                                encoder.shutdown().await?;
+                                Ok(buffer)
+                            }
+                        )*
+                        CompressionType::UNCOMPRESSED => Ok(content.as_ref().to_vec()),
+                    }
+                }
 
-    pub fn convert_stream<T: Stream<Item = io::Result<Bytes>> + Unpin + Send + 'static>(
-        &self,
-        s: T,
-    ) -> Box<dyn Stream<Item = io::Result<Bytes>> + Send + Unpin> {
-        match self {
-            CompressionType::GZIP => {
-                Box::new(ReaderStream::new(GzipDecoder::new(StreamReader::new(s))))
-            }
-            CompressionType::BZIP2 => {
-                Box::new(ReaderStream::new(BzDecoder::new(StreamReader::new(s))))
-            }
-            CompressionType::XZ => {
-                Box::new(ReaderStream::new(XzDecoder::new(StreamReader::new(s))))
+                pub async fn decode(&self, content: impl AsRef<[u8]>) -> io::Result<Vec<u8>> {
+                    match self {
+                        $(
+                            CompressionType::$enum_item => {
+                                let mut buffer = Vec::new();
```

Comment:
```suggestion
                                let mut buffer = Vec::with_capacity(content.len());
```

Reply:
When decode, the content will grow (most compress algorithm may reduce 50%-75% file size in my experience) may be `content.len()*2` is more suitable? (I just wanna say in most case, `Vec::with_capacity(content.len())` still cause vec tor reallocate)

Reply:
Yep, the factor can be extracted as a constant such as 2 or 1.5 etc. It's fine.